hikari/impl/stateful_cache.py

```diff
@@ -1355,55 +1388,52 @@ def _can_remove_member(member: _MemberData, guild_record: _GuildRecord) -> bool:
 
-    def _garbage_collect_member(self, guild_id: snowflake.Snowflake, user_id: snowflake.Snowflake) -> None:
-        guild_record = self._guild_entries.get(guild_id)
-        if guild_record is None or guild_record.members is None or user_id not in guild_record.members:
+    def _garbage_collect_member(self, guild_record: _GuildRecord, member: _MemberData) -> None:
+        if guild_record.members is None or member.id not in guild_record.members:
             return
 
-        member_data = guild_record.members[user_id]
-        if not self._can_remove_member(member_data, guild_record):
+        if not self._can_remove_member(member, guild_record):
             return
 
-        del guild_record.members[user_id]
+        del guild_record.members[member.id]
         if not guild_record.members:
             guild_record.members = None
-            self._remove_guild_record_if_empty(guild_id)
+            self._remove_guild_record_if_empty(member.guild_id)
 
     def _chainable_remove_member(self, member: _MemberData, guild_record: _GuildRecord) -> typing.Optional[_MemberData]:
-        assert guild_record.members
+        assert guild_record.members is not None
         member.has_been_deleted = True
         if not self._can_remove_member(member, guild_record):
             return None
 
         self._garbage_collect_user(member.id, decrement=1)
         del guild_record.members[member.id]
+
+        if not guild_record.members:
+            guild_record.members = None
+            self._remove_guild_record_if_empty(member.guild_id)
+
         return member
 
     def clear_members(
         self,
     ) -> cache.CacheView[snowflake.Snowflake, cache.CacheView[snowflake.Snowflake, guilds.Member]]:
         views = {}
-        cached_users = dict(self._user_entries)
+        cached_users = _copy_mapping(self._user_entries)
 
         def build_member(member: _MemberData) -> guilds.Member:
             return self._build_member(member, cached_users=cached_users)
 
-        for guild_id, guild_record in tuple(self._guild_entries.items()):
+        for guild_id, guild_record in _copy_mapping(self._guild_entries).items():
             if not guild_record.members:
                 continue
 
-            cached_members = {
-                member.id: member
-                for member in itertools.starmap(
-                    self._chainable_remove_member, ((m, guild_record) for m in tuple(guild_record.members.values())),
-                )
-                if member is not None
-            }
-
-            if not guild_record.members:
-                guild_record.members = None
-                self._remove_guild_record_if_empty(guild_id)
-
+            #  This takes roughly half the time a two-layered for loop where we
+            #  assign to the members dict on every inner-iteration takes.
+            members_map = map(
+                lambda m: self._chainable_remove_member(m, guild_record), _copy_mapping(guild_record.members).values(),
```

Comment:
what does chainable do?

Reply:
a chainable method separates logic that's used in multiple view and clear methods during iterations and in some cases allow for chaining stuff within dict literals

Reply:
so like a linked list? 

as opposed to physically call chaing?